phoenix-core/src/main/java/org/apache/phoenix/coprocessor/IndexRebuildRegionScanner.java

```diff
@@ -952,111 +890,89 @@ private void parallelizeIndexVerify(IndexToolVerificationResult.PhaseResult veri
+                break;
+            }
+            indexKey = indexKeyToMutationMap.higherKey(endKeys[regionIndex]);
+            regionIndex++;
         }
-        if (!currentMutationList.isEmpty()) {
-            ungroupedAggregateRegionObserver.checkForRegionClosing();
-            ungroupedAggregateRegionObserver.commitBatchWithRetries(region, currentMutationList, blockingMemstoreSize);
-        }
+        return mapList;
     }
 
-    private void verifyAndOrRebuildIndex() throws IOException {
-        isBeforeRebuilt = true;
-        IndexToolVerificationResult nextVerificationResult = new IndexToolVerificationResult(scan);
-        nextVerificationResult.setScannedDataRowCount(dataKeyToMutationMap.size());
-        if (verifyType == IndexTool.IndexVerifyType.AFTER || verifyType == IndexTool.IndexVerifyType.NONE) {
-            // For these options we start with rebuilding index rows
-            rebuildIndexRows(mutations);
-            nextVerificationResult.setRebuiltIndexRowCount(dataKeyToMutationMap.size());
-            isBeforeRebuilt = false;
-        }
-        if (verifyType == IndexTool.IndexVerifyType.NONE) {
+    private void verifyAndOrRebuildIndex(Map<byte[], List<Mutation>> indexKeyToMutationMap,
+                                         Set<byte[]> mostRecentIndexRowKeys) throws IOException {
+        if (indexKeyToMutationMap.size() == 0) {
             return;
         }
-        if (verifyType == IndexTool.IndexVerifyType.BEFORE || verifyType == IndexTool.IndexVerifyType.BOTH ||
-                verifyType == IndexTool.IndexVerifyType.ONLY) {
-            IndexToolVerificationResult.PhaseResult verificationPhaseResult = new IndexToolVerificationResult.PhaseResult();
-            // For these options we start with verifying index rows
-            parallelizeIndexVerify(verificationPhaseResult);
-            nextVerificationResult.getBefore().add(verificationPhaseResult);
-        }
-        if (verifyType == IndexTool.IndexVerifyType.BEFORE || verifyType == IndexTool.IndexVerifyType.BOTH) {
-            // For these options, we have identified the rows to be rebuilt and now need to rebuild them
-            // At this point, dataKeyToDataPutMap includes mapping only for the rows to be rebuilt
-            mutations.clear();
-
-            for (Map.Entry<byte[], Pair<Put, Delete>> entry: dataKeyToMutationMap.entrySet()) {
-                if (entry.getValue().getFirst() != null) {
-                    mutations.add(entry.getValue().getFirst());
-                }
-                if (entry.getValue().getSecond() != null) {
-                    mutations.add(entry.getValue().getSecond());
-                }
+        if (verifyType != IndexTool.IndexVerifyType.ONLY) {
+            // Since we are building the index table, the region boundaries may change while we are building. Thus,
+            // we should update endKeys
+            try (org.apache.hadoop.hbase.client.Connection connection =
```

Comment:
Creating HBase connections is heavy-weight. Can we re-use one from somewhere, such as a Phoenix connection?

Also, since this code is duplicated in the constructor, can we factor out into a small helper function?

Reply:
What are the implications of endKeys being out-of-date? Is it a correctness problem, or a perf problem (as tasks would be crossing region boundaries)

Reply:
As I mentioned above, I will fix this, i.e., I  will use one connection per table region.  I will also look into sharing the same connection within a region server or within a cluster. I think there is a way to do this. @abhishek-chouhan, let me know if you know already there is a way to do this.